### ISSUE TITLE
Environment not passed to criu when starting container from checkpoint

### ISSUE THREAD

**Issue by tswift242** [docker/docker#30031]
**Description**

I'm not sure if this is a docker, containerd, or runc issue, but it seems that at least one of these layers is not passing the current environment to criu when a container is started from a checkpoint (probably also when creating a checkpoint). Criu has certain options that can be set through environment variables, and these env vars don't seem to be respected when I do C/R through docker. For example, the CRIU_FAULT env var is not being respected, which we need to maintain good restore performance when we do many restores on a single host. See this criu issue for more info: https://github.com/xemul/criu/issues/268#issuecomment-271553587

**Steps to reproduce the issue:**
1. export CRIU_FAULT=130
2. docker run --security-opt=seccomp:unconfined --name cr -d ubuntu /bin/bash -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 5; done'
3. docker checkpoint create cr test_cp
4. docker start --checkpoint test_cp cr

**Describe the results you received:**

Did not see "Force no-breakpoints restore" (https://github.com/xemul/criu/blob/a7faf367a6b2e215a86f2fffe785c63aa3902c96/criu/parasite-syscall.c#L1432) in criu restore log, indicating CRIU_FAULT=130 setting wasn't respected.

**Describe the results you expected:**

Should see "Force no-breakpoints restore" in criu restore log, indicating CRIU_FAULT=130 setting is respected.

**Output of `docker version`:**

```
Client:
 Version:      1.13.0-rc5
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 03:07:30 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-rc5
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 03:07:30 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 46
 Running: 46
 Paused: 0
 Stopped: 0
Images: 27
Server Version: 1.13.0-rc5
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 51371867a01c467f08af739783b8beafc154c4d7
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.39-040439-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 120.1 GiB
Name: ip-10-97-2-120
ID: NSKT:GWC5:CRQH:BBVT:CSWF:BNL6:6T7O:AMSA:TAM6:72OO:CHOK:INU4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 425
 Goroutines: 342
 System Time: 2017-01-10T18:34:33.989626544Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

AWS

**Comment by thaJeztah** [comment]
ping @boucher @xemul ptal

**Comment by boucher** [comment]
How does Docker deal with environment settings in other contexts? (Or has this never come up?) 

The environment of your current shell is not surprisingly not going to apply to the environment the actual `criu` process is running in: the cli is just making calls to the daemon, which has its own environment, which in turn makes calls to containerd, runc, then criu. The last two layers are actual process calls at least, and have some opportunity for inheriting the env based on settings.

**Comment by thaJeztah** [comment]
Yes, getting those environment variables from the host probably isn't gonna work (also keeping in mind that the client could run on a whole different machine than the daemon is running). Theoretically, it would be possible to get all `CRIU_` env vars from the current environment when running `docker checkpoint create` (and pass those on to `criu`?), but I think options should be passed through a flag, if they are useful to have. I'm not sure to what extend we should go to expose all CRIU options.

Does criu have a way to specify the same options as an argument / flag instead of just environment variables? Thinking of `docker checkpoint create --opt fault=130 .....`, which would pass those options on to criu.

**Comment by boucher** [comment]
CRIU has dozens of flags, and there was some concern that exposing them all
at the Docker level would be too confusing (and too tied to CRIU
specifically).

**Comment by thaJeztah** [comment]
Agreed, _if_ there are flags that are needed, it should be a curated list of options, and only if they are "generic" enough

**Comment by tswift242** [comment]
Right, so the whole crux of this issue is how to use the docker-criu integration with non-default criu options (non-default from the perspective of which options docker always passes to criu). It looks like criu may be trying to come up with a better configuration system on its end (https://github.com/xemul/criu/issues/278), in which case maybe docker doesn't have to do anything.

criu has a lot of different options, so I agree that docker shouldn't have to have a flag for each one. I think a generic mechanism (e.g. a config file) for passing non-default criu options, either on the criu side or the docker side, is important to have though.

**Comment by xemul** [comment]
One thing that bothers me about xemul/criu#278 is -- there are (roughly) two types of options to CRIU. First are those that describe some fundamental configuration, like maximum size of unlinked files to be taken into image, which port to use by page-server, where to put logs and irmap cache and alike. These option can be put in a node-wide config files. But there are options, that only makes sense here and now. E.g. -- whether or not to C/R TCP connection or the map of external bind mounts. These options if put into a node-wide config file will affect all running checkpoints in a way caller of those _others_ checkpoints might not want to have. So having some way to specify option for a _particular_ invocation of CRIU might still be needed.

**Comment by zhouxinyu** [comment]
Hi @xemul , Do we have any update on this issue? I have no idea that how to pass arguments from docker command to criu, eg:

```
#docker checkpoint create --tcp-established --checkpoint-dir=/tmp --leave-running priceless_stonebraker checkpoint2
unknown flag: --tcp-established
See 'docker checkpoint create --help'.
```
